### PR TITLE
nll experiment: compute SCCs instead of iterative region solving

### PR DESCRIPTION
**Pull request by nikomatsakis**
This is an attempt to speed up region solving by replacing the current iterative dataflow with a SCC computation. The idea is to detect cycles (SCCs) amongst region constraints and then compute just one value per cycle. The graph with all cycles removed is of course a DAG, so we can then solve constraints "bottom up" once the liveness values are known.

I kinda ran out of time this morning so the last commit is a bit sloppy but I wanted to get this posted, let travis run on it, and maybe do a perf run, before I clean it up.

### CODE REVIEW

**Comment by nikomatsakis** [comment]
Perf URL for when data becomes available:

http://perf.rust-lang.org/compare.html?start=4faaf7e3359fa78bad2e8c54011e94ce8ac078c6&end=3fcccbf34a7af46ec467dc21c321db20994357ea&stat=instructions%3Au

**Comment by Eh2406** [comment]
That link does not seem to be working.

**Comment by nikomatsakis** [comment]
@Eh2406 yeah something went wrong with perf :( probably we will have results in about 2.5 hours from now

**Comment by nikomatsakis** [comment]
OK, apparently the perf results are available at least partially now. This doesn't appear to be a big win. I'm not sure why, I'd like to dig a little more (I'd also like to land at least *some* of these refactorings regardless...).

**Comment by Eh2406** [comment]
The link still is not working for me. How do you find the perf results you referred to?

**Comment by Eh2406** [comment]
Link is now working, numbers look pretty good. double digit nll improvements in `inflate-check` `inflate-debug` `inflate-opt` but a 17% regression in `clap-rs-check`. I wonder why?

**Comment by nikomatsakis** [comment]
For some reason, scc generation takes 10% of clap-rs time.

**Comment by nikomatsakis** [comment]
Er, wait, that may be incorrect.

**Comment by nikomatsakis** [comment]
OK, I think I found the problem -- I wasn't checking for duplicates. I am rebasing now, but local measurements suggest that this wipes region solving off the profile.

**Comment by nikomatsakis** [comment]
Comparison URL for later:

http://perf.rust-lang.org/compare.html?start=8dd715ee5e462384668e3c83c17ee641e9776b64&end=8e6d8db2bf93a7c311c0a7a4cc7e1b4b7a5a5c04&stat=instructions%3Au

**Comment by nikomatsakis** [comment]
So the perf results for *this* branch are available, but not the ideal comparison point.

- [Dashboard for this PR](https://perf.rust-lang.org/nll-dashboard.html?commit=8e6d8db2bf93a7c311c0a7a4cc7e1b4b7a5a5c04&stat=instructions%3Au)
- [Comparison for `master^`, which ought to be reasonable](https://perf.rust-lang.org/compare.html?start=a739c51d108b1958a1c3e145588035c580e7973a&end=8e6d8db2bf93a7c311c0a7a4cc7e1b4b7a5a5c04&stat=instructions%3Au)

**Comment by nikomatsakis** [comment]
OK, so, this is now a massive win for inflate (45%) without any real downside (but not much benefit either). It also simplifies the code quite a bit, in terms of reasoning about inference, so I think we ought to land it.

**Comment by nikomatsakis** [comment]
@Eh2406 I mean that it doesn't benefit the *other tests* -- it's just neutral.

**Comment by nikomatsakis** [comment]
OK, so, I sorta wanted to cleanup the "one master commit" but I'm feeling lazy now, so instead I just documented in the commit message the various things that it does. I think this is ready for review.

**Comment by nikomatsakis** [comment]
That said, @michaelwoerister is probably not the right person — for one thing, not sure if they are back yet from leave. I think @pnkfelix will be around though so:

r? @pnkfelix

**Comment by nikomatsakis** [comment]
Er, actually, @pnkfelix may not be back. I'm going to make it

r? @eddyb

**Comment by michaelwoerister** [comment]
I'm back but I don't know this part of the code.

**Comment by pnkfelix** [comment]
(i'm back and will get on reviewing this asap)

**Comment by nikomatsakis** [comment]
@pnkfelix I rebased -- this required a few tweaks because I needed to make the constraint graph persist for error reporting (before, it was created and then discarded once SCCs were computed) -- but nothing too major. Hopefully it'll pass tests, I didn't have time to run them to completion.

File: src/librustc/mir/mod.rs [hunk 1]
```diff
@@ -22,7 +22,7 @@ use mir::visit::MirVisitable;
 use rustc_apfloat::ieee::{Double, Single};
 use rustc_apfloat::Float;
 use rustc_data_structures::control_flow_graph::dominators::{dominators, Dominators};
-use rustc_data_structures::control_flow_graph::ControlFlowGraph;
+use rustc_data_structures::control_flow_graph;
```
Comment (pnkfelix) [inline src/librustc/mir/mod.rs]: ultra-nit: shouldn't this just be a `self` added to the `use` on the the next line?

File: src/librustc_data_structures/graph/scc/mod.rs [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+
+    /// The stack of successors: as we visit a node, we mark our
+    /// position in this stack, and when we encounter a successor SCC,
+    /// we push it on the stack. When we complete an SCC, we can pop
+    /// everything off the stack that was found along the way.
+    successors_stack: Vec<S>,
+
+    /// A set used to strip duplicates. As we accumulate successors
+    /// into the successors_stack, we sometimes get duplicate entries.
+    /// We use this set to remove those -- we keep it around between
```
Comment (pnkfelix) [inline src/librustc_data_structures/graph/scc/mod.rs]: I would rephrase "we keep it around" to make it clearer that you are not actually reusing the previously computed duplicate set. Maybe just "we keep its storage around" ?

File: src/librustc_data_structures/graph/scc/mod.rs [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+    /// is that we do a depth-first search. Along the way, we keep a
+    /// stack of each node whose successors are being visited. We
+    /// track the depth of each node on this stack (there is no depth
+    /// if the node is not on the stack). When we find that some node
+    /// N with depth D can reach some other node N' with lower depth
+    /// D' (i.e., D' < D), we know that N, N', and all nodes in
+    /// between them on the stack are part of an SCC.
+    ///
+    /// For each node, we track the lowest depth of any successor we
+    /// have found, along with that
```
Comment (pnkfelix) [inline src/librustc_data_structures/graph/scc/mod.rs]: along with that ... what?

File: src/librustc_data_structures/graph/scc/test.rs [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+    assert_eq!(sccs.scc(2), 0);
+    assert_eq!(sccs.scc(3), 2);
+    assert_eq!(sccs.successors(0), &[]);
+    assert_eq!(sccs.successors(1), &[0]);
+    assert_eq!(sccs.successors(2), &[0]);
+}
+
+#[test]
+fn test_find_state_2() {
+    // The order in which things will be visited is important to this
```
Comment (pnkfelix) [inline src/librustc_data_structures/graph/scc/test.rs]: hmm. I might put the comment with the graph *above* this comment here.  Its a tough call; on the one hand, it makes sense to try to keep the ascii art close to the actual code for the graph's definition. On the other hand, I really would like to know what the graph is before I read this comment...

(Maybe the right thing would be to move the ascii art *and* the `let graph = ...;` to up above this comment that starts with "The order in which things will be visited is ...")

**Comment by pnkfelix** [comment]
The commit with a log message that starts with "compute region values using SCCs not iterative flow" has a typo, it says: "we compute SCCs once all outlives constraints are not", but "not" cannot be the right word there. ("Known"? "Frozen"?)

File: src/librustc_data_structures/graph/scc/mod.rs [hunk 5]
```diff
@@ -54,6 +54,11 @@ impl<N: Idx, S: Idx> Sccs<N, S> {
         self.scc_data.len()
     }
 
+    /// Returns the number of SCCs in the graph.
```
Comment (pnkfelix) [inline src/librustc_data_structures/graph/scc/mod.rs]: (You should either delete this comment or revise it to something like "returns an iterator over the SCCs in the graph. Right now it just distracts, IMO...)

Comment (pnkfelix) [reply src/librustc_data_structures/graph/scc/mod.rs]: ((or even just "iterate over the SCCs in the graph", I guess...))

File: src/librustc_mir/borrow_check/nll/region_infer/error_reporting/mod.rs [hunk 6]
```diff
@@ -316,7 +308,7 @@ impl<'tcx> RegionInferenceContext<'tcx> {
 
         while changed {
             changed = false;
-            for constraint in &*self.constraints {
+            for constraint in self.constraints.iter() {
                 if let Some(n) = result_set[constraint.sup] {
```
Comment (pnkfelix) [inline src/librustc_mir/borrow_check/nll/region_infer/error_reporting/mod.rs]: Elsewhere, in `fn find_constraint_paths_between_regions_helper`, you have changed the direction of the `constraint` edge so that `'a: 'b` is represented as an edge `'a -> 'b` instead of `'b -> 'a`, which in effect meant that you replaced occurrences of `.sup` with `.sub`. Do you need to similarly revise the code in this method, swapping  `.sup` with `.sub`  and vice versa?

(In other words, I'm not sure I follow why the meaning of "'a may depend on 'b" seems to have changed in this commit. Unless the case in  `fn find_constraint_paths_between_regions_helper` was simply buggy?)

File: src/test/ui/nll/closure-requirements/propagate-approximated-fail-no-postdom.stderr [hunk 7]
```diff
@@ -1,28 +1,28 @@
...
    |
 LL |         |_outlives1, _outlives2, _outlives3, x, y| {
    |          ----------              ---------- lifetime `'2` appears in this argument
    |          |
    |          lifetime `'1` appears in this argument
-...
-LL |             demand_y(x, y, p) //~ ERROR
-   |             ^^^^^^^^^^^^^^^^^ argument requires that `'1` must outlive `'2`
+LL |             // Only works if 'x: 'y:
+LL |             let p = x.get(); //~ ERROR
```
Comment (pnkfelix) [inline src/test/ui/nll/closure-requirements/propagate-approximated-fail-no-postdom.stderr]: At first blush I was going to ask whether we were better off with the previous diagnostic, where the call to `demand_y(..)` is highlighted, since in my mind the relevant constraint on the relationship between `'x` and `'y` is "introduced" by that call, and so it is arguably the important thing to highlight.

But after some review of the test and further thought, I think I've convinced myself that this new diagnostic is also good, and perhaps better, since it is telling us where the '`x` is flowing into a variable that will eventually need to satisfy `'y`.

So at this point I'm more just writing this comment to ask you whether you have any thoughts on the relative value of the two spans. Should we be trying to highlight both of them in some manner, in the long term, perhaps?

**Comment by pnkfelix** [comment]
Okay I finished my review.

Almost all the code seemed fine to me. The only part that I was really squinting at was the potential issue I pointed out [here](https://github.com/rust-lang/rust/pull/51987#discussion_r202034709), but maybe that is "just fine". It wasn't immediately obvious so I figured it was worth asking rather than puzzling over.

So you can r=me at your pleasure.

File: src/librustc_mir/borrow_check/nll/region_infer/error_reporting/mod.rs (hunk 6 above)
Comment (nikomatsakis) [reply src/librustc_mir/borrow_check/nll/region_infer/error_reporting/mod.rs]: Hmm, so I didn't touch this code, and it operates as it ever did. The problem here is that this code really ought to be removed -- it's an older variant that `find_constraint_paths_between_regions` has superceded, but it is still used in one or two places -- it'd probably be worth making an effort to remove it entirely, but I'm inclined to open an issue on that and defer it to another PR.

File: src/test/ui/nll/closure-requirements/propagate-approximated-fail-no-postdom.stderr (hunk 7 above)
Comment (nikomatsakis) [reply src/test/ui/nll/closure-requirements/propagate-approximated-fail-no-postdom.stderr]: Eventually, I had planned to highlight both spans, yes. I wanted to go through and label all the "significant" points along the path, basically -- but right now we just pick one (and apparently our choice is somewhat influenced here by the order of the search).

File: src/librustc_data_structures/graph/scc/mod.rs (hunk 3 above)
Comment (nikomatsakis) [reply src/librustc_data_structures/graph/scc/mod.rs]: heh I think I got bored of trying to write up the algorithm...

**Comment by nikomatsakis** [comment]
@bors retry

I'm gonna hazard a guess that this is something transient:

> [00:29:25] /rustroot/bin/ld:/checkout/obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers:1397: syntax error in VERSION script

cc @rust-lang/infra

**Comment by pietroalbini** [comment]
Yeah, it's a known problem -- https://github.com/rust-lang/rust/issues/50887

**Comment by nnethercote** [comment]
This was a 38% NLL win for `inflate`, but and 8.5% NLL loss for `html5ever` (already by far the worst NLL benchmark), and a ~1% NLL loss for a bunch more benchmarks:
https://perf.rust-lang.org/compare.html?start=c0955a34bcb17f0b31d7b86522a520ebe7fa93ac&end=bce32b532de61434841b7c2ce3085e1f63d6a7a1&stat=instructions:u

@nikomatsakis: Was that expected?

**Comment by nnethercote** [comment]
It also improved `style-servo` by 4% and `clap-rs` by 2%.
